components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/cache/BaseCache.java

```diff
@@ -0,0 +1,396 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.caching.impl.CacheImpl;
+import org.wso2.carbon.caching.impl.CachingConstants;
+import org.wso2.carbon.context.CarbonContext;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.identity.organization.management.service.util.OrganizationManagementConfigUtil;
+import org.wso2.carbon.identity.organization.management.service.util.Utils;
+
+import java.io.Serializable;
+import java.util.concurrent.TimeUnit;
+
+import javax.cache.Cache;
+import javax.cache.CacheBuilder;
+import javax.cache.CacheConfiguration;
+import javax.cache.CacheManager;
+import javax.cache.Caching;
+
+/**
+ * A base class for all organization management cache implementations.
+ * Duplicate of org.wso2.carbon.identity.core.cache.BaseCache class.
+ *
+ * @param <K> cache key type.
+ * @param <V> cache value type.
+ */
+public abstract class BaseCache<K extends Serializable, V extends Serializable> {
+
+    private static final Log log = LogFactory.getLog(BaseCache.class);
+    private static final String CACHE_MANAGER_NAME = "OrganizationManagementCacheManager";
+    private CacheBuilder<K, V> cacheBuilder;
+    private String cacheName;
+    private final OrgMgtCacheConfig orgMgtCacheConfig;
+
+    public BaseCache(String cacheName) {
+
+        this(cacheName, false);
+    }
+
+    public BaseCache(String cacheName, boolean isTemp) {
+
+        this.cacheName = CachingConstants.LOCAL_CACHE_PREFIX + cacheName;
+        orgMgtCacheConfig = OrganizationManagementConfigUtil.getOrgMgtCacheConfig(CACHE_MANAGER_NAME, cacheName);
+        if (orgMgtCacheConfig != null) {
+            if (orgMgtCacheConfig.isDistributed()) {
+                this.cacheName = cacheName;
+            }
+            orgMgtCacheConfig.setTemporary(isTemp);
+        }
+        if (log.isDebugEnabled()) {
+            String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            log.debug("Cache : " + cacheName + "  is initialized for tenant domain : " + tenantDomain);
+        }
+    }
+
+    private synchronized Cache<K, V> getBaseCache() {
+
+        Cache<K, V> cache;
+        CacheManager cacheManager = Caching.getCacheManagerFactory()
+                .getCacheManager(CACHE_MANAGER_NAME);
+
+        if (cacheBuilder == null) {
+            cacheManager.removeCache(cacheName);
+            cacheBuilder = cacheManager.<K, V>createCacheBuilder(cacheName).
+                    setExpiry(CacheConfiguration.ExpiryType.ACCESSED,
+                            new CacheConfiguration.Duration(TimeUnit.SECONDS, getCacheTimeout())).
+                    setExpiry(CacheConfiguration.ExpiryType.MODIFIED,
+                            new CacheConfiguration.Duration(TimeUnit.SECONDS, getCacheTimeout())).
+                    setStoreByValue(false);
+            cache = cacheBuilder.build();
```

Comment:
saw a code where we can capture a concurrent exception - https://www.tabnine.com/code/java/classes/javax.cache.CacheManager

Better check

Reply:
Reverted the improvements done on the BaseCache and suppressed the double checked locking warning with a todo item on addressing it properly. https://github.com/wso2-extensions/identity-organization-management-core/pull/51/commits/66ac04b4e682566222dd6264884f5c1c3935a443